internal/cluster/cluster_service.go

```diff
@@ -0,0 +1,329 @@
+		doneCh:            make(chan struct{}),
+		startCh:           make(chan struct{}, 1),
+		invocationFactory: bundle.InvocationFactory,
+		eventDispatcher:   bundle.EventDispatcher,
+		logger:            bundle.Logger,
+		membersMap:        newMembersMap(),
+		config:            bundle.Config,
+	}
+	service.ownerConnectionAddr.Store(&pubcluster.AddressImpl{})
+	return service
+}
+
+func (s *ServiceImpl) GetMemberByUUID(uuid string) pubcluster.Member {
+	return s.membersMap.Find(uuid)
+}
+
+func (s *ServiceImpl) Start(wantSmartRouting bool) <-chan struct{} {
+	subscriptionID := event.MakeSubscriptionID(s.handleOwnerConnectionOpened)
+	s.eventDispatcher.Subscribe(EventOwnerConnectionChanged, subscriptionID, s.handleOwnerConnectionOpened)
+	s.eventDispatcher.Subscribe(EventMembersUpdated, event.DefaultSubscriptionID, s.handleMembersUpdated)
+	if wantSmartRouting {
+		s.listenPartitionsLoaded()
+	}
+	go s.logStatus()
+	return s.startCh
+}
+
+func (s *ServiceImpl) Stop() {
+	subscriptionID := event.MakeSubscriptionID(s.handleOwnerConnectionOpened)
+	s.eventDispatcher.Unsubscribe(EventOwnerConnectionChanged, subscriptionID)
+	close(s.doneCh)
+}
+
+func (s *ServiceImpl) SmartRoutingEnabled() bool {
+	return atomic.LoadInt32(&s.smartRouting) == smartRoutingEnabled
+}
+
+func (s *ServiceImpl) MemberAddrs() []string {
+	return s.membersMap.MemberAddrs()
+}
+
+func (s *ServiceImpl) memberCandidateAddrs() []*pubcluster.AddressImpl {
+	addrSet := NewAddrSet()
+	for _, addrProvider := range s.addrProviders {
+		addrSet.AddAddrs(addrProvider.Addresses())
+	}
+	return addrSet.Addrs()
+}
+
+func (s *ServiceImpl) handleOwnerConnectionOpened(event event.Event) {
+	if ownerConnectionOpenedEvent, ok := event.(*OwnerConnectionChanged); ok {
+		go s.sendMemberListViewRequest(ownerConnectionOpenedEvent.Conn)
+	}
+}
+
+func (s *ServiceImpl) handleMembersUpdated(event event.Event) {
+	if membersUpdateEvent, ok := event.(*MembersUpdated); ok {
+		added, removed := s.membersMap.Update(membersUpdateEvent.Members, membersUpdateEvent.Version)
+		if atomic.CompareAndSwapInt32(&s.startChAtom, 0, 1) {
+			close(s.startCh)
+			s.startCh = nil
+		}
+		if len(added) > 0 {
+			s.eventDispatcher.Publish(NewMembersAdded(added))
+		}
+		if len(removed) > 0 {
+			s.eventDispatcher.Publish(NewMemberRemoved(removed))
+		}
+	}
+}
+
+func (s *ServiceImpl) sendMemberListViewRequest(conn *Connection) {
+	request := codec.EncodeClientAddClusterViewListenerRequest()
+	inv := s.invocationFactory.NewConnectionBoundInvocation(request, -1, nil, conn, s.config.InvocationTimeout, func(response *proto.ClientMessage) {
+		codec.HandleClientAddClusterViewListener(response, func(version int32, memberInfos []pubcluster.MemberInfo) {
+			s.logger.Debug(func() string { return "members updated" })
+			s.eventDispatcher.Publish(NewMembersUpdated(memberInfos, version))
+		}, func(version int32, partitions []proto.Pair) {
+			s.eventDispatcher.Publish(NewPartitionsUpdated(partitions, version))
```

Comment:
Why are we not making a call to directly to "partition_service.handlePartitionsUpdated" ?
Is there a reason for this to go over an event dispatcher ?

Reply:
Updated at: https://github.com/yuce/hazelcast-go-client/commit/15136dbdd26d8f9e1acdf9eba6c5ef62be6392d4